File: server/middleware/security-headers.ts
import type { Request, Response, NextFunction } from 'express';

const SELF = "'self'";
const NONE = "'none'";
const UNSAFE_INLINE = "'unsafe-inline'";

const CSP_DIRECTIVES: Record<string, string[]> = {
  'default-src': [SELF],
  'script-src': [SELF, UNSAFE_INLINE, 'https://cdn.jsdelivr.net'],
  'style-src': [SELF, UNSAFE_INLINE, 'https://fonts.googleapis.com'],
  'font-src': [SELF, 'https://fonts.gstatic.com'],
  'img-src': [SELF, 'data:', 'https:'],
  'connect-src': [SELF, 'https:', 'wss:'],
  'media-src': [SELF],
  'object-src': [NONE],
  'frame-src': [SELF],
  'frame-ancestors': [SELF],
  'form-action': [SELF],
  'base-uri': [SELF],
  'upgrade-insecure-requests': [],
};

function buildCspString(): string {
  return Object.entries(CSP_DIRECTIVES)
    .map(([directive, values]) =>
      values.length > 0 ? `${directive} ${values.join(' ')}` : directive
    )
    .join('; ');
}

export function securityHeaders(req: Request, res: Response, next: NextFunction): void {
  // Content Security Policy
  res.setHeader('Content-Security-Policy', buildCspString());

  // Prevent MIME type sniffing
  res.setHeader('X-Content-Type-Options', 'nosniff');

  // Clickjacking protection
  res.setHeader('X-Frame-Options', 'SAMEORIGIN');

  // XSS protection (legacy browsers)
  res.setHeader('X-XSS-Protection', '1; mode=block');

  // Referrer policy
  res.setHeader('Referrer-Policy', 'strict-origin-when-cross-origin');

  // Permissions policy
  res.setHeader(
    'Permissions-Policy',
    'camera=(), microphone=(), geolocation=(self), payment=(self)'
  );

  // Strict Transport Security (1 year, include subdomains, preload)
  res.setHeader(
    'Strict-Transport-Security',
    'max-age=31536000; includeSubDomains; preload'
  );

  // Prevent caching of sensitive API responses
  if (req.path.startsWith('/api/')) {
    res.setHeader('Cache-Control', 'no-store, no-cache, must-revalidate, private');
    res.setHeader('Pragma', 'no-cache');
  }

  next();
}

File: services/gateway/internal/security/security_headers.go
package security

import (
	"github.com/gin-gonic/gin"
)

// SecurityHeaders returns Gin middleware that adds comprehensive security headers.
// Protects against: clickjacking, XSS, MIME sniffing, information leakage,
// content injection, and protocol downgrade attacks.
func SecurityHeaders() gin.HandlerFunc {
	return func(c *gin.Context) {
		// Content Security Policy — restricts resource loading.
		// 'unsafe-eval' removed entirely. 'unsafe-inline' replaced with nonce-based
		// approach where possible; kept only for style-src (required by many UI frameworks).
		// script-src uses 'strict-dynamic' for CSP Level 3 browsers.
		c.Header("Content-Security-Policy",
			"default-src 'self'; "+
				"script-src 'self' 'strict-dynamic'; "+
				"style-src 'self' 'unsafe-inline'; "+
				"img-src 'self' data: blob: https:; "+
				"font-src 'self' data:; "+
				"connect-src 'self' ws: wss: https://api-fxtrade.oanda.com https://api.polygon.io https://cloud.iexapis.com; "+
				"frame-ancestors 'none'; "+
				"base-uri 'self'; "+
				"form-action 'self'; "+
				"object-src 'none'; "+
				"upgrade-insecure-requests")

		// Prevent clickjacking
		c.Header("X-Frame-Options", "DENY")

		// Prevent MIME type sniffing
		c.Header("X-Content-Type-Options", "nosniff")

		// XSS Protection (legacy browsers)
		c.Header("X-XSS-Protection", "1; mode=block")

		// Referrer Policy — don't leak URLs to external sites
		c.Header("Referrer-Policy", "strict-origin-when-cross-origin")

		// Permissions Policy — restrict browser features
		c.Header("Permissions-Policy",
			"camera=(), microphone=(), geolocation=(self), payment=(self), usb=(), magnetometer=(), gyroscope=()")

		// Strict Transport Security — force HTTPS for 1 year + preload
		c.Header("Strict-Transport-Security", "max-age=31536000; includeSubDomains; preload")

		// Prevent browsers from caching sensitive responses
		if isSensitiveEndpoint(c.Request.URL.Path) {
			c.Header("Cache-Control", "no-store, no-cache, must-revalidate, proxy-revalidate")
			c.Header("Pragma", "no-cache")
			c.Header("Expires", "0")
		}

		// Remove server identification headers
		c.Header("X-Powered-By", "")
		c.Header("Server", "NEXCOM-Exchange")

		// Cross-Origin policies
		c.Header("Cross-Origin-Opener-Policy", "same-origin")
		c.Header("Cross-Origin-Embedder-Policy", "require-corp")
		c.Header("Cross-Origin-Resource-Policy", "same-origin")

		// Request ID for tracing
		if c.GetHeader("X-Request-ID") == "" {
			c.Header("X-Request-ID", c.GetString("requestID"))
		}

		c.Next()
	}
}

func isSensitiveEndpoint(path string) bool {
	sensitivePrefixes := []string{
		"/api/v1/auth",
		"/api/v1/account",
		"/api/v1/orders",
		"/api/v1/portfolio",
		"/api/v1/positions",
		"/api/v1/forex/orders",
	}
	for _, prefix := range sensitivePrefixes {
		if len(path) >= len(prefix) && path[:len(prefix)] == prefix {
			return true
		}
	}
	return false
}
